File: lib/services/skill_router.dart
import 'package:powerful_students/models/study_skill.dart';

/// Seleziona la skill più pertinente in base al messaggio utente (keyword scoring).
class SkillRouter {
  SkillRouter._();

  static const idNone = 'none';

  /// Restituisce l'id skill con punteggio più alto, o [idNone] se nessun match.
  static String route(String message, List<StudySkill> skills) {
    final normalized = _normalize(message);
    if (normalized.isEmpty) {
      return idNone;
    }

    var bestId = idNone;
    var bestScore = 0;

    for (final skill in skills) {
      if (skill.id == idNone || skill.routingKeywords.isEmpty) {
        continue;
      }

      var score = 0;
      for (final keyword in skill.routingKeywords) {
        final k = _normalize(keyword);
        if (k.isEmpty) {
          continue;
        }
        if (normalized.contains(k)) {
          // Frasi lunghe pesano di più delle singole parole corte.
          score += k.length >= 8 ? 4 : 2;
        }
      }

      if (score > bestScore) {
        bestScore = score;
        bestId = skill.id;
      }
    }

    return bestScore > 0 ? bestId : idNone;
  }

  static String _normalize(String text) {
    return text.toLowerCase().replaceAll(RegExp(r'\s+'), ' ').trim();
  }
}

File: ai-playground/lib/skill-router.ts
import type { SkillEntry } from './skills';

function normalize(text: string): string {
  return text.toLowerCase().replace(/\s+/g, ' ').trim();
}

/** Seleziona la skill con punteggio keyword più alto, o `none`. */
export function routeSkill(message: string, skills: SkillEntry[]): string {
  const normalized = normalize(message);
  if (!normalized) {
    return 'none';
  }

  let bestId = 'none';
  let bestScore = 0;

  for (const skill of skills) {
    if (skill.id === 'none' || !skill.keywords?.length) {
      continue;
    }

    let score = 0;
    for (const keyword of skill.keywords) {
      const k = normalize(keyword);
      if (!k) {
        continue;
      }
      if (normalized.includes(k)) {
        score += k.length >= 8 ? 4 : 2;
      }
    }

    if (score > bestScore) {
      bestScore = score;
      bestId = skill.id;
    }
  }

  return bestScore > 0 ? bestId : 'none';
}
